File: 프로그래머스/unrated/133500. 등대/등대.java
import java.util.ArrayList;

class Solution {
     private final int[][] DP = new int[2][100001];
        private ArrayList<ArrayList<Integer>> Adj;

        void go(int node) {
            DP[1][node] = 1;

            for (int child : Adj.get(node)) {
                if (DP[1][child] == 0) {
                    go(child);
                    DP[0][node] += DP[1][child];
                    DP[1][node] += Math.min(DP[1][child], DP[0][child]);
                }
            }
        }

        public int solution(int n, int[][] lighthouse) {
            Adj = new ArrayList<>();
            for (int i = 0; i <= n; i++)
                Adj.add(new ArrayList<>());

            for (int[] house : lighthouse) {
                int a = house[0], b = house[1];

                Adj.get(a).add(b);
                Adj.get(b).add(a);
            }

            go(1);

            return Math.min(DP[1][1], DP[0][1]);
        }
}

File: 프로그래머스/unrated/133500. 등대/등대.cpp
#include <string>
#include <vector>
#include <iostream>
#include <queue>
#include <cmath>
#include <algorithm>

#define MAX 100001
#define INF 987654321

using namespace std;

vector<int> Adj[MAX];
bool Visited[MAX];
int DP[2][MAX];

void go(int node) {
    DP[0][node] = 0, DP[1][node] = 1;
    Visited[node] = true;

    for (auto &next: Adj[node]) {
        if (Visited[next])
            continue;

        go(next);
        DP[0][node] += DP[1][next];
        DP[1][node] += min(DP[1][next], DP[0][next]);
    }
}

int solution(int n, vector<vector<int>> lighthouse) {
    for (auto &l: lighthouse) {
        Adj[l[0]].push_back(l[1]);
        Adj[l[1]].push_back(l[0]);
    }

    go(1);
    
    return min(DP[0][1], DP[1][1]);
}